Diy2

nitinks-dnz opened the pull request.

nitinks-dnz committed:
Added pool and routines functionality with test cases

nitinks-dnz committed:
Resolved channel termination issue

nitinks-dnz committed:
Resolved ProcessWithExpiry going on infinite loop issue

nitinks-dnz committed:
Updated the test cases

nitinks-dnz committed:
Resolved ProcessWithExpiry going on infinite loop issue

nitinks-dnz committed:
Added logic to control CPUs to be used by services and updated test cases

nitinks-dnz committed:
Increasing test coverage

nitinks-dnz committed:
Refactoring

nitinks-dnz committed:
Adding the Dockerfile to run the test cases in container

nitinks-dnz committed:
Code refactoring and adding the comments

nitinks-dnz committed:
Update ReadeMe

sammithk commented:
Please try to do the following

-  Use only **buffered channels** to limit maximum number of go routines.
-  Use channels for synchronising across go routines.
-  Use **concurrency/parallelism** wherever possible to minimise latency.
-  Use **sync.Once** for creating a singleton library.

sammithk commented on pool.go:
```diff
@@ -0,0 +1,174 @@
+)
+
+// Pool is a struct which contains the list of routines
+type Pool struct {
```
Please make data members publicly available.

nitinks-dnz replied:
Updated Pool data members public. Ref- [Commit 99f5cd](https://github.com/nitinks-dnz/go-pool/commit/99f5cdd7efe22cd61b3a27e921a960194beb828a)

nitinks-dnz committed:
Updating Formatting in README

nitinks-dnz committed:
Using sync.Once to instantiate pool as singleton

nitinks-dnz committed:
Using buffered channels to limit max go routines, Usage of concurrency to minimize latency, removal of unwanted operations

nitinks-dnz committed:
Updating Pool data members public

nitinks-dnz commented:
1. Updated the initialisation to be used as singleton library. Ref - [Commit ab29130](https://github.com/nitinks-dnz/go-pool/commit/ab29130b90e2cf406563fd6513cf6358201bed21)
2. Removed the usage of slice for go routines and its dependant functionalities, replaces with usage of buffered channel to limit number of go routines. Ref - [Commit e3ece11](https://github.com/nitinks-dnz/go-pool/commit/e3ece11e539b29d234281670af98316604dfd3a2)
3. Added parallelism to minimize latency. Ref - [Commit e3ece11](https://github.com/nitinks-dnz/go-pool/commit/e3ece11e539b29d234281670af98316604dfd3a2)
4. Updated documentation with proper formatting. Ref - [Commit 32b938d](https://github.com/nitinks-dnz/go-pool/pull/1/commits/32b938d6602454b26bcd24bcbd984dcfcbb1ede9)

sammithk commented on pool.go:
```diff
@@ -0,0 +1,124 @@
+
+	retPayload, retOk := <-p.RetChan
+	if !retOk {
+		panic(ErrWorkerClosed)
```
why panic, why not return error.

nitinks-dnz replied:
Removed Panic and throwing the proper error for the cause. Ref - [Commit 6debada](https://github.com/nitinks-dnz/go-pool/pull/1/commits/6debadacb7e0141868462ef5c953f08e99e001e6)

sammithk commented on pool.go:
```diff
@@ -0,0 +1,124 @@
+	var wg sync.WaitGroup
+	for i := 0; i < n; i++ {
+		wg.Add(1)
+		go p.worker(&wg, p.WorkerFun())
```
Why spawn all workers at the same time, why not dynamically spawn them when required and make sure max active go routines are capped at n.

nitinks-dnz replied:
Dynamically spawning the go routines as and when required by maintaining the counter. Ref - [Commit 1a3db2c](https://github.com/nitinks-dnz/go-pool/pull/1/commits/1a3db2c0366f77d53520f51be49b7be6704041a2)

nitinks-dnz committed:
Dynamically spawning go routines

nitinks-dnz committed:
Returning the error and removing Panic

nitinks-dnz committed:
Added model and Handling of custom tasks sent by clients

nitinks-dnz committed:
Controlling the maximum number of go routines using buffered channels
